### PR TITLE
Orchagent: Integrate P4Orch

### PR DESCRIPTION
**Pull request by donNewtonAlpha**
* support pushing switch configurations.
    * support ACL table definitions.
    * support insert/delete/modify in APPL_DB.
    * support for L3 forwarding, WCMP, and ACL tables.
    * support for response path through APPL_STATE_DB.

    Co-authored-by: Robert Halstead rhalstea@google.com
    Co-authored-by: Srikishen Pondicherry Shanmugam kishanps@google.com
    Co-authored-by: Steffen Smolka smolkaj@google.com
    Co-authored-by: Jonathan Dilorenzo dilo@google.com
    Co-authored-by: Richard Yu richardyu@google.com
    Co-authored-by: Brian O'Connor bocon@opennetworking.org
    Co-authored-by: Konstantin Weitz konne@google.com
    Co-authored-by: Stefan Heule heule@google.com
    Co-authored-by: Tianyu Xia tyxia@google.com

**What I did**

**Why I did it**

**How I verified it**

**Details if related**

### CODE REVIEW

**Comment by donNewtonAlpha** [comment]
@StephenWangGoogle

File: orchagent/acltable.h [hunk 1]
```diff
@@ -59,3 +61,28 @@ typedef enum
 } acl_table_type_t;
 
 typedef std::unordered_map<std::string, acl_table_type_t> acl_table_type_lookup_t;
+typedef std::map<std::string, sai_acl_stage_t> acl_stage_lookup_t;
+typedef std::map<sai_acl_stage_t, sai_switch_attr_t> acl_stage_to_switch_attr_lookup_t;
+
+struct AclTableGroupMember {
```
Comment (prsunny) [inline orchagent/acltable.h:67]: Please move to {-> new line and for other structs

File: orchagent/orchdaemon.cpp [hunk 2]
```diff
@@ -589,6 +592,10 @@ bool OrchDaemon::init()
 
     m_orchList.push_back(&CounterCheckOrch::getInstance(m_configDb));
 
+    vector<string> p4rt_tables = {APP_P4RT_TABLE_NAME};
```
Comment (prsunny) [inline orchagent/orchdaemon.cpp:595]: Should this be done based on global flag? `gResponsePublisherRecord`

File: orchagent/switchorch.cpp [hunk 3]
```diff
@@ -55,6 +59,157 @@ SwitchOrch::SwitchOrch(DBConnector *db, vector<TableConnector>& connectors, Tabl
     querySwitchTpidCapability();
     auto executorT = new ExecutableTimer(m_sensorsPollerTimer, this, "ASIC_SENSORS_POLL_TIMER");
     Orch::addExecutor(executorT);
+    initAclGroupsBindToSwitch();
```
Comment (prsunny) [inline orchagent/switchorch.cpp:62]: As discussed in work-group meeting, lets move this to `p4orch` initialization

File: tests/dvslib/dvs_acl.py [hunk 4]
```diff
@@ -152,7 +152,12 @@ def get_acl_table_group_ids(self, expected: int) -> List[str]:
         Returns:
             The list of ACL group IDs in ASIC DB.
         """
-        acl_table_groups = self.asic_db.wait_for_n_keys(self.ADB_ACL_GROUP_TABLE_NAME, expected)
+        num_keys = len(self.asic_db.default_acl_groups) + expected
```
Comment (prsunny) [inline tests/dvslib/dvs_acl.py:155]: please revert the test changes

File: orchagent/acltable.h (hunk 1 above)
Comment (bocon13) [reply orchagent/acltable.h]: Fixed

File: orchagent/switchorch.cpp (hunk 3 above)
Comment (bocon13) [reply orchagent/switchorch.cpp]: Fixed

File: tests/dvslib/dvs_acl.py (hunk 4 above)
Comment (bocon13) [reply tests/dvslib/dvs_acl.py]: Done

File: orchagent/orchdaemon.cpp (hunk 2 above)
Comment (bocon13) [reply orchagent/orchdaemon.cpp:603]: @baxia-lan @mint570 what do you think?

Comment (mint570) [reply orchagent/orchdaemon.cpp:603]: gResponsePublisherRecord is to enable/disable the APPL STATE DB record file. I would suggest to use a different flag to enable/disable p4orch. Brian, is it possible to pass the p4orch enable flag from build time? If not, we can just hardcode one in the code (similar to gResponsePublisherRecord)?

Comment (baxia-lan) [reply orchagent/orchdaemon.cpp:603]: We do not need to have a flag here. Initialization of p4orch will not impact anything, it just listens to APP_P4RT_TABLE_NAME("P4RT" table in APP DB). As long as no entry is added in the"P4RT" table in APP DB, P4Orch will do nothing.

Comment (prsunny) [reply orchagent/orchdaemon.cpp:603]: I think yes, we probably don't need a flag here. Just make sure the p4orch init/constructor is not invoking something that a legacy platform does not support. Like an example below, where we are moving `initAclGroupsBindToSwitch`  to p4orch.

Comment (bocon13) [reply orchagent/orchdaemon.cpp:603]: It should be safe to call the constructor. `initAclGroupsBindToSwitch` isn't called until a P4 program is pushed.

**Comment by rlhui** [comment]
@donNewtonAlpha - please get all tests passed, seek help from @prsunny  if needed.

**Comment by bocon13** [comment]
Rebased to include #1997 and retesting

**Comment by bocon13** [comment]
Resolved merge conflicts with #1951

**Comment by mint570** [comment]
@prsunny 
I am not sure why the BuildArm is not finished. But it looks like all builds are passing.
There is one alert on copy constructor. Just style alert.
I think we are good to go.

**Review by prsunny — APPROVED** [review]

**Comment by prsunny** [comment]
@mint570 , could you please update submodule?

**Comment by mint570** [comment]
> @mint570 , could you please update submodule?

https://github.com/Azure/sonic-buildimage/pull/9417

**Pull request merged**